### PR TITLE
RES: take into account namespace during shadowing

### LINKED ISSUE #4544: Unable to resolve anyhow's `context` with `anyhow!`
## Environment

* **IntelliJ Rust plugin version:** 0.2.107.2137-192
* **Rust toolchain version:** rustc 1.37.0 (eae3437df 2019-08-13)
* **IDE name and version:** IntelliJ IDEA 2019.2.3 (Ultimate Edition)
Build #IU-192.6817.14, built on September 24, 2019
* **Operating system:** macOS

## Steps to reproduce

Make a new project, use this as main.rs

```rust
use anyhow::{anyhow, Context, Result as AnyResult};
use std::collections::HashMap;

fn main() -> AnyResult<()> {
    println!("Hello, world!");

    let mut data: HashMap<String, String> = HashMap::new();
    data.insert("foo".to_string(), "bar".to_string());
    data.get("thing").context("no thing in hash")?;

    Err(anyhow!("failed anyway"))
}
```

context will be undefined

Screen Shot 2019-10-19 at 11 47 21 AM

But this works:

```rust
use anyhow::{Context, Result as AnyResult};
use std::collections::HashMap;

fn main() -> AnyResult<()> {
    println!("Hello, world!");

    let mut data: HashMap<String, String> = HashMap::new();
    data.insert("foo".to_string(), "bar".to_string());
    data.get("thing").context("no thing in hash")?;

    Ok(())
}
```

So perhaps it's a dupe of https://github.com/intellij-rust/intellij-rust/issues/4509 though the symptom doesn't seem as explosive as what's described there.

I can still resolve (and click-navigate) other things like `anyhow!`, `AnyResult`, etc.

### PR DESCRIPTION
**Pull request by vlad20012**
Fixes: #4509, fixes #4544

### CODE REVIEW

File: src/main/kotlin/org/rust/lang/core/resolve/Processors.kt [hunk 1]
```diff
@@ -28,6 +28,7 @@ interface ScopeEntry {
     val name: String
     val element: RsElement?
     val subst: Substitution get() = emptySubstitution
+    val isNotLazyOrInitialized: Boolean get() = true
```
Comment (Undin) [inline src/main/kotlin/org/rust/lang/core/resolve/Processors.kt:31]: Maybe just `isInitialized` it enough? Non lazy things are initialized by default

File: src/main/kotlin/org/rust/lang/core/resolve/NameResolution.kt [hunk 2]
```diff
@@ -1471,21 +1471,41 @@ private tailrec fun PsiFile.unwrapCodeFragments(): PsiFile {
...
-        e.name !in prevScope && run {
-            currScope += e.name
-            processor(e)
+        val prevNs = prevScope[e.name]
+        (prevNs == null || (e.element as? RsNamedElement)?.namespaces?.intersects(prevNs) != true) && run {
+            val result = processor(e)
+            if (e.isInitialized) {
+                val newNs = (e.element as? RsNamedElement)?.namespaces
+                if (newNs != null) {
+                    currScope[e.name] = prevNs?.let { it + newNs } ?: newNs
```
[the remark this answers is not in the document]
Comment (vlad20012) [reply src/main/kotlin/org/rust/lang/core/resolve/NameResolution.kt:1486]: `prevNs.orEmpty() + newNs` allocates new set. I don't think it's reasonable in such a hot place

File: src/main/kotlin/org/rust/lang/core/resolve/NameResolution.kt [hunk 3]
```diff
@@ -1431,10 +1432,9 @@ fun processNestedScopesUpwards(
 
     val prelude = findPrelude(scopeStart)
     if (prelude != null) {
-        // XXX: fix prelude items resolve on the nightly. Revert it to `v -> v.name !in prevScope`
```
[the remark this answers is not in the document]
Comment (vlad20012) [reply src/main/kotlin/org/rust/lang/core/resolve/NameResolution.kt:1434]: It was fixed a long time ago with `cfg` attributes support =)

File: src/main/kotlin/org/rust/lang/core/resolve/NameResolution.kt [hunk 4]
```diff
@@ -1471,21 +1471,41 @@ private tailrec fun PsiFile.unwrapCodeFragments(): PsiFile {
...
+    }
+}
+
+inline fun processWithShadowingImmutable(
+    prevScope: Map<String, Set<Namespace>>,
+    crossinline processor: RsResolveProcessor,
+    f: (RsResolveProcessor) -> Boolean
+): Boolean {
+    return f { e: ScopeEntry ->
+        val prevNs = prevScope[e.name]
```
[the remark this answers is not in the document]
Comment (vlad20012) [reply src/main/kotlin/org/rust/lang/core/resolve/NameResolution.kt:1505]: Well, I don't think it's too complicated. And I think this lambda-based `processIfNotShadowed` is more complicated =)

[the remark this answers is not in the document]
Comment (Undin) [reply src/main/kotlin/org/rust/lang/core/resolve/NameResolution.kt:1505]: I don't construction `expr && run { do something complex }` very much. Especially, when `expr` is quite complex too (in our case it is `prevNs != null && (e.element as? RsNamedElement)?.namespaces?.intersects(prevNs) == true`)
Can we avoid them?

File: src/main/kotlin/org/rust/lang/core/resolve/NameResolution.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (Undin) [reply src/main/kotlin/org/rust/lang/core/resolve/NameResolution.kt:1434]: Ok

File: src/main/kotlin/org/rust/lang/core/resolve/NameResolution.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (vlad20012) [reply src/main/kotlin/org/rust/lang/core/resolve/NameResolution.kt:1506]: I've extracted a part of condition to a separate `if` and removed `run`

[the remark this answers is not in the document]
Comment (Undin) [reply src/main/kotlin/org/rust/lang/core/resolve/NameResolution.kt:1506]: Thanks!

**Comment by vlad20012** [comment]
bors r=undin

**Pull request merged**